### DISCUSSION TITLE
mfront installation and use - CAST3M interface

### DISCUSSION THREAD

**Discussion by RonanRiverie** [thelfer/tfel#457] [Q&A] [answered]
Hello,

I am Ronan, the PhD student and Deborah contacted you about some helps to use mfront. I am just starting with this so I have very basic questions.

Since I have understood that it was quite complex to install it on Windows, I installed Ubuntu on VirtualBox in order to be able to use mfront together with CAST3M.

First of all, I tried to follow the tutorial from this page : https://tfel.sourceforge.net/install.html#sec:QuickUbuntu and find some already answered questions on this forum but I did not managed to install properly mfront. I probably did something wrong since I am getting an error. So there there are probably some libraries missing.

![Screenshot from 2023-12-06 15-52-23](https://github.com/thelfer/tfel/assets/65529689/2d84caf4-16a8-4a6f-a87e-de8fe62401da)

After I installed CAST3M on Ubuntu but I am still getting the same error. I also installed CAST3M on Windows and it is not clear if mfront is already integrated in CAST3M and how to use it ? I mean with what command should I launch mfront since it is not a recognized command. on Windows. And on Linux, should mfront be automatically present when installing CAST3M (I installed the developer and mfront CAST3M version during the installation).

For the moment I am just trying to make it work like using the norton example. After, I would like to use the EKMC model to study strain aging (which has already been implemented in mfront thanks to this page : https://github.com/thelfer/tfel/discussions/316 but after I would like to try implement it by myself).

Thank you very much for your help to make mfront work.

**Comment by thelfer** [comment]
@RonanRiverie Is there any reason not to use the `MFront` version delivered with `Cast3M` ? It is not required, but much simplier.

If you do wish to use the lastest version, let us move on !

For you problem, is is just a matter of updating properly `LD_LIBRARY_PATH` for the linker to find the `TFEL` libraries. Here is the script that I use:

```sh
export TFELHOME=/home/th202608/codes/tfel/master/install-python-3.7
export PATH=$TFELHOME/bin:$PATH
export LD_LIBRARY_PATH=$TFELHOME/lib:$LD_LIBRARY_PATH
export PYTHONPATH=$TFELHOME/lib/python3.7/site-packages/:$PYTHONPATH
```

You must change the first line to match your installation path. The last line is not required if you did not compile the `python` bindings.

I save this script in a file named `env.sh` that I place at the root of `TFEL` installation directory.  Then each time I want to use `MFront`, I do:

~~~~
source /home/th202608/codes/tfel/master/install-python-3.7/env.sh
~~~~

That allows me to have several versions in //.

**Reply by RonanRiverie** [reply]
Thanks a lot for your fast answer.

Yes actually it would be even better to use it directly in CAST3M but I don't understand how to use it so, that's why I was starting trying to install it directly.

I am going to try your lines.

**Reply by thelfer** [reply]
To use the `MFront` version of `Cast3M`, you can do:

```bash
$ compilcast23 file.mfront
```

This scripts sets the environment variables correctly and calls `MFront` behind the scene.

If you want to use `MFront` directly (which is more convenient) or `MTest`, you can use my previous `env.sh` with 

~~~~
TFELHOME=<path_to_castem>/MFRONT
~~~~

In my case:

~~~~
export TFELHOME=/home/th202608/codes/castem/2022/install/MFRONT
export PATH=$TFELHOME/bin:$PATH
export LD_LIBRARY_PATH=$TFELHOME/lib:$LD_LIBRARY_PATH
~~~~

**Reply by RonanRiverie** [reply]
Perfect, the first line works well:
![image](https://github.com/thelfer/tfel/assets/65529689/34a604fc-c92c-4d5e-b834-8f2656a111a0)

However, not for the second or maybe it is not the right way to do it (see screenshot).

**Reply by RonanRiverie** [reply]
The first line even works on Windows:

![image](https://github.com/thelfer/tfel/assets/65529689/c20a0610-2209-4c9e-aca5-ba182e04ca22)

**Reply by thelfer** [reply]
ok, so `compilcast23` works. 

Now, the script:

1. You shall "source" it once per terminal. You don't have to call it multiple times
2. Can you give me the outputs of (after sourcing `env.sh`):

~~~~
$ echo $LD_LIBRARY_PATH
~~~~

**Reply by RonanRiverie** [reply]
Here is the output of the command:

![image](https://github.com/thelfer/tfel/assets/65529689/8a3e4b6c-0307-4c3c-9682-a313bba6b72f)

So it seems ok ?

**Reply by thelfer** [reply]
Yes.

But `norton.mfront` is not a command. You shall type `mfront norton.mfront`

**Reply by RonanRiverie** [reply]
Ah ok, thanks but I am getting the same problem as without CAST3M (I hope both mfront are not getting in conflict explaining why I am getting an error) :

![image](https://github.com/thelfer/tfel/assets/65529689/69df9e45-96d3-492a-ba07-c77ade5dcffe)

**Reply by thelfer** [reply]
You seems to have mess up things. The fact that `mfront` requires `libTFELMFront.so.5.0.0-dev` shows that there something related to the `master` version.
You shall start from scratch in a new terminal. Then check that the following commands are consistent:

- `echo $LD_LIBRARY_PATH`
- `which mfront`

**Reply by RonanRiverie** [reply]
Ok so I indeed messed up something (the make install installation step):
The issue is first I installed in the wrong location: /usr/local so I reinstalled it in /home/ronan/Documents/TFEL but it seems to still be directed in the first location:

![image](https://github.com/thelfer/tfel/assets/65529689/ba3c9528-fff7-42e9-945a-57c12fe78ae4)

**Reply by thelfer** [reply]
Indeed, installing in `/usr/local` was a very bad idea.
I don't even know if `cmake` provides a `make uninstall` command...
You may need to remove every files by hand

**Reply by RonanRiverie** [reply]
Yes I forgot to change the location in cmake...
So I will start everything from scratch and I will let you know if it works better.
Anyway, thanks a lot for your help !

**Reply by thelfer** [reply]
Good luck !

**Reply by RonanRiverie** [reply]
Ok, so I started everything aging from the right beginning and now, this is working fine with the norton.mfront and norton.mtest examples:

![image](https://github.com/thelfer/tfel/assets/65529689/8a5fd98e-d6ef-462e-ba62-5e562c2fdc62)

Thank you very much Thomas for your help :)

**Reply by RonanRiverie** [reply]
And just one last question:
So compilcast23 also works on Windows but how to launch the mtest file on Windows?

**Reply by thelfer** [reply]
I give a course today on windows, i'll surely get more information

**Reply by RonanRiverie** [reply]
Ok nice, So I am going to continue getting familiar with mfront on Linux and let me know if you have more information on Windows.

**Reply by thelfer** [reply, marked as answer]
For windows, you have to open the command line (`cmd.exe`) and type:

```
set PATH=c:\Cast3M\PCW_23\GCC\bin;%PATH%
set PATH=c:\Cast3M\PCW_23\MFRONT\bin;%PATH%
set TFELHOME=c:\Cast3M\PCW_23\MFRONT
set MAKE=mingw32-make.exe
```

You have to check that the paths are correct. Check also the name of the make executable which is located at `c:\Cast3M\PCW_23\GCC\bin`

**Reply by RonanRiverie** [reply]
It works perfectly fine !

![image](https://github.com/thelfer/tfel/assets/65529689/6d69d5fe-f19d-4101-aa3b-d360abd1df54)


Thanks a lot !

**Reply by thelfer** [reply]
Not really, convergence is awfull :)

When using the `Cast3M` interface, the default is to use the elastic stiffness as the tangent operator. You may want to add an acceleration algorithm, as follows:

```python
@AccelerationAlgorithm 'Cast3M';
```

**Reply by RonanRiverie** [reply]
Ah yes indeed, so I added this line in the mtest file and now I get 856 iterations instead of 14930.
So, the acceleration algorithm changes the way the tangent operator is computed?

**Reply by thelfer** [reply]
No, not a all. See https://www.sciencedirect.com/science/article/pii/S0898122115004046 for details :)

**Reply by RonanRiverie** [reply]
Ah ok, I will have a look at it.

**Comment by maxence-wz** [comment]
Hi Ronan,
MFront seems to be installed but not sourced correctly. You need to add the folder containing libTFELMFront.so.5.0.0-dev to your LD_LIBRARY_PATH with the command :
export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:directoryPath
(where directoryPath is the absolute path containing the libTFELMFront.so.5.0.0-dev library)
Let me know if it works,
Maxence

**Reply by thelfer** [reply]
@RonanRiverie This is more or less my anwser :) Thanks @maxence-wz

**Reply by RonanRiverie** [reply]
So I tried both solutions but I probably did not do it correctly since I am getting the same error: 

![image](https://github.com/thelfer/tfel/assets/65529689/55a0d5f0-ddc7-4f59-84e5-6d82ed624680)

I installed everything in /home/ronan/Documents and tried to put the env.sh file directly in Documents or in /home/ronan/Documents/TFEL or in /home/ronan/Documents/TFEL/tfel

**Reply by thelfer** [reply]
What is the output of:

~~~~
$ echo $LD_LIBRARY_PATH
~~~~

**Reply by RonanRiverie** [reply]
![image](https://github.com/thelfer/tfel/assets/65529689/613ebe34-b715-44a7-9416-358f846a169f)

So I might probably not be in the right folder with "TFELHOME" since I don't have the /lib folder no ?

**Reply by thelfer** [reply]
Probably !

**Reply by thelfer** [reply]
`mfront` is installed in `$TFELHOME/bin`. That may help !

**Reply by thelfer** [reply]
There seems to be a folder `tfel` for the sources and `build` for building the project, but where is the `install` folder ?

Did you make `make install` ?

**Reply by RonanRiverie** [reply]
I think I did it in that folder, yes :

![image](https://github.com/thelfer/tfel/assets/65529689/c0ae609d-b4dc-475f-9b9f-8b4d4c0d1de3)

**Reply by RonanRiverie** [reply]
But maybe not since I don't have the install folder, maybe I can try again.

**Reply by thelfer** [reply]
That's the sources. There is no folders `bin`, `lib`, `share`, etc...
